### PR TITLE
fix postgres autoIncrement seq with createIfNoExists issue #2099

### LINKED ISSUE #2099: `createIfNotExists` doesn't create auto increment sequences
**DESCRIPTION:**

Using `createIfNotExists` does not create auto increment sequences, whereas `create` works just fine in creating those sequences.

A symptom of this issue is if you use `createIfNotExists` with `O.AutoInc` fields, you will get these errors:

```
postgres    | 2020-02-15 13:46:18.160 UTC [412] ERROR:  null value in column "id" violates not-null constraint
```

This doesn't happen if you have the correct sequence.

**STEPS TO REPRODUCE:**

Clone [mdibaiee/slick-bug-sequences](https://github.com/mdibaiee/slick-bug-sequences) and run the code as is initially:
```
createdb slick_bug

DATABASE_URL=postgresql://testuser:testpassword@localhost:5432/slick_bug sbt run
```

**EXPECTED RESULT:**

Sequence should be created and no error should be thrown.

**ACTUAL RESULT:**

You will get an error:
```
[error] (run-main-0) org.postgresql.util.PSQLException: ERROR: null value in column "id" violates not-null constraint
[error]   Detail: Failing row contains (null, Test).
[error] org.postgresql.util.PSQLException: ERROR: null value in column "id" violates not-null constraint
[error]   Detail: Failing row contains (null, Test).
[error]         at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)
[error]         at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2183)
[error]         at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:308)
[error]         at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
[error]         at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
[error]         at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:143)
[error]         at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:120)
[error]         at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$SingleInsertAction.$anonfun$run$15(JdbcActionComponent.scala:520)
[error]         at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:425)
[error]         at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement$(JdbcBackend.scala:420)
[error]         at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:489)
[error]         at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:511)
[error]         at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$SingleInsertAction.run(JdbcActionComponent.scala:517)
[error]         at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)
[error]         at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)
[error]         at slick.dbio.DBIOAction$$anon$4.$anonfun$run$3(DBIOAction.scala:239)
[error]         at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
[error]         at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
[error]         at scala.collection.AbstractIterable.foreach(Iterable.scala:921)
[error]         at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:239)
[error]         at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:237)
[error]         at slick.basic.BasicBackend$DatabaseDef$$anon$3.liftedTree1$1(BasicBackend.scala:276)
[error]         at slick.basic.BasicBackend$DatabaseDef$$anon$3.run(BasicBackend.scala:276)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error]         at java.base/java.lang.Thread.run(Thread.java:835)
```

Also try connecting to the database (`psql slick_bug`) and check for sequences and tables:

```
\dt
\ds
```

You will see tables existing, but no sequences.

Now try changing the `src/main/scala/example/Hello.scala` file and replacing `createIfNotExists` with `create`. Drop the database and recreate and run again:

```
dropdb slick_bug
createdb slick_bug

DATABASE_URL=postgresql://testuser:testpassword@localhost:5432/slick_bug sbt run
```

This time there will be no errors and you can check that the sequence is created successfully in postgres shell (`psql slick_bug`):

```
\ds
```

**SETUP:**

Libraries:
```
  "com.typesafe.slick" %% "slick" % "3.3.2",
  "org.postgresql" % "postgresql" % "42.2.5" //org.postgresql.ds.PGSimpleDataSource dependency
```

Postgres 12.1

### PR DESCRIPTION
**Pull request by mdibaiee**
Fixes #2099

### CODE REVIEW

**Comment by jfilali** [comment]
Any plan to merge this? It seems pretty straightforward but will prevent users to spend enormous amounts of time trying to figure out why it doesn't work as expected.   Thank so much.

**Comment by SethTisue** [comment]
if this is still relevant, rebase on current master? then we can try again to find a reviewer

**Comment by mdibaiee** [comment]
Hey @SethTisue, I just rebased on master and pushed again. 👍

**Comment by SethTisue** [comment]
Would anyone watching this repo like to review it?

(It looks this was already discussed pretty thoroughly on the bug, so if no new reviewer appears in the next week or two, I'd say we could merge this anyway.)

**Comment by nafg** [comment]
Closing and reopening to trigger CI

**Review by nafg — DISMISSED** [review]

**Comment by nafg** [comment]
This works, but it seems more like a workaround rather than a proper fix. If you look all the way back to where `autoIncrement = false` was introduced, 805c05a14133d7627dbfc110bf5a1f6a52f44405, it seems like it was just a lazy shortcut to avoid overriding `appendOptions` like was done for other drivers, to just get it not to append `AUTO_INCREMENT` since `SERIAL` was specified as the type. I think `autoIncrement` should be kept as `true` and the postgres ddl should just override `appendOptions` to be the same as in basic ddl except without AUTO_INCREMENT.

**Review by nafg — CHANGES_REQUESTED** [review]
Can you supply a test, and try my suggestion of overriding `appendOptions` and never reassigning `autoIncrement`?

File: slick/src/main/scala/slick/jdbc/PostgresProfile.scala [hunk 1]
```diff
@@ -238,13 +238,20 @@ trait PostgresProfile extends JdbcProfile {
...
+    }
+
     override def appendColumn(sb: StringBuilder): Unit = {
       sb append quoteIdentifier(column.name) append ' '
       if(autoIncrement && !customSqlType) {
         sb append (if(sqlType.toUpperCase == "BIGINT") "BIGSERIAL" else "SERIAL")
       } else appendType(sb)
-      autoIncrement = false
       appendOptions(sb)
+      autoIncrement = originalAutoIncrement
```
Comment (nafg) [inline slick/src/main/scala/slick/jdbc/PostgresProfile.scala:254]:
```suggestion
```

**Review by nafg — CHANGES_REQUESTED** [review]

**Comment by nafg** [comment]
@mdibaiee are you planning on taking this further?

**Comment by mdibaiee** [comment]
Hey @nafg, sorry for late replies. I still have not got to test this new approach. A minimal test case is available here though, I just need to run it: https://github.com/mdibaiee/slick-bug-sequences

**Comment by mdibaiee** [comment]
I did a manual test with the example here and it worked! https://github.com/mdibaiee/slick-bug-sequences

I will see if I can write an automated test now 👍

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by mdibaiee** [comment]
On my local instance the test does not run for postgres apparently 🤔 
I'm not sure if I can make it run for postgres as well, or if the pipeline is enough for that?

**Review by nafg — APPROVED** [review]

**Comment by nafg** [comment]
Failure: https://github.com/slick/slick/pull/2100/checks?check_run_id=2907117315#step:9:2039

Seems it like fails only on MySQL, with "Incorrect table definition; there can be only one auto column and it must be defined as a key."

```
[error] Test com.typesafe.slick.testkit.tests.JdbcMiscTest.testAutoIncrementCreateIfNotExists[mysql] failed: java.sql.SQLSyntaxErrorException: Incorrect table definition; there can be only one auto column and it must be defined as a key, took 0.039 sec
[error]     at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:120)
[error]     at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
[error]     at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:953)
[error]     at com.mysql.cj.jdbc.ClientPreparedStatement.execute(ClientPreparedStatement.java:370)
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$6.$anonfun$run$7(JdbcActionComponent.scala:292)
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$6.$anonfun$run$7$adapted(JdbcActionComponent.scala:292)
[error]     at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:425)
[error]     at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement$(JdbcBackend.scala:420)
[error]     at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:489)
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$6.$anonfun$run$6(JdbcActionComponent.scala:292)
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$6.$anonfun$run$6$adapted(JdbcActionComponent.scala:292)
[error]     at scala.collection.immutable.Vector.foreach(Vector.scala:1856)
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$6.run(JdbcActionComponent.scala:292)
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$6.run(JdbcActionComponent.scala:290)
[error]     at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)
[error]     at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)
[error]     at slick.dbio.DBIOAction$$anon$4.$anonfun$run$3(DBIOAction.scala:239)
[error]     at scala.collection.immutable.Vector.foreach(Vector.scala:1856)
[error]     at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:239)
[error]     at slick.dbio.DBIOAction$$anon$4.run(DBIOAction.scala:237)
[error]     at slick.basic.BasicBackend$DatabaseDef$$anon$3.liftedTree1$1(BasicBackend.scala:276)
[error]     at slick.basic.BasicBackend$DatabaseDef$$anon$3.run(BasicBackend.scala:276)
[error]     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error]     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error]     at java.lang.Thread.run(Thread.java:829)
```

**Comment by mdibaiee** [comment]
@nafg ah, so I think I need to make that column explicitly a key for this test to pass on MySQL. I will look at it 👍

**Review by nafg — APPROVED** [review]

**Pull request merged**